### ISSUE TITLE
Support for Chrome > 50 push messages with encrypted payload data

### ISSUE THREAD

**Issue by grri** [Redth/PushSharp#726]
I am able to send notifications via PushSharp and GCM to Chrome but currently without payload data.

Does PushSharp currently support the encrypted payload data for web push notifications? described here - https://developers.google.com/web/updates/2016/03/web-push-encryption

If not, are you planning to implement the Encryption?

cheers

**Comment by gbrussa** [comment]
That would be very nice. I've tried to compute the ECDH with .Net Cryptography and also with BouncyCastle but didn't succeed in either case

**Comment by elcapitanandres** [comment]
gbrussa  "I've tried to compute the ECDH with .Net Cryptography and also with BouncyCastle but didn't succeed in either case"
We are two now...

**Comment by LbISS** [comment]
Yep, we need it! :)

**Comment by star1ite** [comment]
Yes, I'm waiting too. Here is java version for this: https://github.com/web-push-libs/webpush-java/tree/master/src/main/java/nl/martijndwars/webpush

**Comment by LbISS** [comment]
C# version:
http://stackoverflow.com/a/39839330/1758830

**Comment by bm-software** [comment]
@LbISS Thank you for the link, I've tested it but unfortunately it doesn't works :( 
At this line:
`var IKM = agreementGenerator.CalculateAgreement(new ECPublicKeyParameters(spec.Curve.DecodePoint(userKey), spec));`

It throws the exception:
`Invalid point encoding 66 `

Any solution?

**Comment by nGoline** [comment]
@bm-software I'm trying to use the helper but i get the same `Invalid point encoding 66` error as you but I could not overcome it.
I was not able to follow the instructions on the link because the `WebEncoders.Base64UrlDecode` method returns a byte[] and the substitute `Base64ForUrlDecode` returns a string. Even though I could not get nothing but `null` from `byte[] decbuff = HttpServerUtility.UrlTokenDecode(str);`.
Has anyone got a little farther with this?

**Comment by PashaPash** [comment]
@nGoline got chrome push encryption working based on https://github.com/LogicSoftware/WebPushEncryption.

Check https://github.com/LogicSoftware/PushSharp/commits/web_push ([see diff with master](https://github.com/Redth/PushSharp/compare/master...LogicSoftware:web_push)) for complete integration code.

**Comment by dherbe** [comment]
Thanks @PashaPash.
Also based on https://github.com/LogicSoftware/WebPushEncryption I've managed to implement PushWebNotification.
I've added the PushSharp.Web namespace to my project and I easily set the broker configuration and queuing just like I already had done for Android and iOS.

What I had to change the most was the client side registration of the EndPoint (already had it) but now the Keys (P256dh, Auth). As mentioned here https://developers.google.com/web/updates/2016/03/web-push-encryption.
